File: csharp/parsing-log-files/ParsingLogFiles.cs
using System;
using System.Text.RegularExpressions;

public class LogParser
{
    public bool IsValidLine(string text)
    {
        return Regex.IsMatch(text, @"^\[TRC\].*|^\[DBG\].*|^\[INF\].*|^\[WRN\].*|^\[ERR\].*|^\[FTL\].*");
    }

    public string[] SplitLogLine(string text)
    {
        string pattern = @"\<[\^\*\=\-]*\>";
        return Regex.Split(text, pattern);
    }

    public int CountQuotedPasswords(string lines)
    {
        int count = 0;
        string[] allLines = lines.Split(Environment.NewLine);
        foreach (string line in allLines)
        {
            if (Regex.IsMatch(line, "\\\".*password.*\\\"", RegexOptions.IgnoreCase))
            {
                count++;
            }
        }
        return count;
    }

    public string RemoveEndOfLineText(string line)
    {
        return Regex.Replace(line, @"end-of-line\d*", "");
    }

    public string[] ListLinesWithPasswords(string[] lines)
    {
        throw new NotImplementedException($"Please implement the LogParser.ListLinesWithPasswords() method");
    }
}

File: go/parsing-log-files/parsing_log_files.go
package parsinglogfiles

import "regexp"

func IsValidLine(text string) bool {
	var re = regexp.MustCompile(`^\[(TRC|DBG|INF|WRN|ERR|FTL)\]`)
	return re.MatchString(text)
}

func SplitLogLine(text string) []string {
	re := regexp.MustCompile(`<[*~=-]*>`)
	return re.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	var result int
	re := regexp.MustCompile(`\".*(?i)password.*\"`)
	for _, line := range lines {
		if re.MatchString(line) {
			result++
		}
	}
	return result
}

func RemoveEndOfLineText(text string) string {
	re := regexp.MustCompile(`end-of-line\d*`)
	return re.ReplaceAllString(text, "")
}

func TagWithUserName(lines []string) []string {
	re := regexp.MustCompile(`User\s+(\w+)`)
	for i, line := range lines {
		match := (re.FindStringSubmatch(line))
		if len(match) > 1 {
			lines[i] = "[USR] " + match[1] + " " + line
		}
	}
	return lines
}
